### PR TITLE
fix: resolve extern prelude for local mods in block modules

### PR DESCRIPTION
**Pull request by roife**
fix https://github.com/rust-lang/rust-analyzer/issues/17057, https://github.com/rust-lang/rust-analyzer/issues/17032.

We should use `ModuleOrigin` to check if the current module is a pseudo-module introduced by blocks (where names might be shadowed), rather than checking `block_def_map`.

### CODE REVIEW

**Comment by roife** [comment]
I'm not sure how to add unittests for it. 🤔

File: crates/hir-def/src/nameres/path_resolution.rs [hunk 1]
```diff
@@ -470,9 +472,9 @@ impl DefMap {
         };
 
         let extern_prelude = || {
-            if self.block.is_some() {
-                // Don't resolve extern prelude in block `DefMap`s, defer it to the crate def map so
-                // that blocks can properly shadow them
+            if matches!(self[module].origin, ModuleOrigin::BlockExpr { .. }) {
```
Comment (Veykril) [inline crates/hir-def/src/nameres/path_resolution.rs:475]:
```suggestion
            if self.block.is_some() && module == DefMap::ROOT {
```

If I understand the code here correctly, likjewise below. Though I am unsure whether this is the correct fix here

Comment (Veykril) [reply crates/hir-def/src/nameres/path_resolution.rs:475]: Though it should be more correct than the status quo

Comment (roife) [reply crates/hir-def/src/nameres/path_resolution.rs:475]: They are equivalent. If a module is a `BlockExpr`, its `local_id` should be equivalent to `ROOT`, and it should be within a block; conversely, this is also true. Should I change it to the latter?

Comment (Veykril) [reply crates/hir-def/src/nameres/path_resolution.rs:475]: Yes, sorry I wasn't clear. With unsure whether this is the correct fix I meant the general approach here. My code suggestion is the more correct way to do your check here (they are equivalent, it's just that the origin shouldn't be used to check this property).

File: crates/hir-def/src/nameres/path_resolution.rs [hunk 2]
```diff
@@ -515,8 +518,8 @@ impl DefMap {
             None => self[Self::ROOT].scope.get(name),
         };
         let from_extern_prelude = || {
-            if self.block.is_some() {
-                // Don't resolve extern prelude in block `DefMap`s.
+            if matches!(self[module].origin, ModuleOrigin::BlockExpr { .. }) {
+                // Don't resolve extern prelude in pseudo-module of a block.
```
Comment (Veykril) [inline crates/hir-def/src/nameres/path_resolution.rs:522]: Same thing here

**Comment by Veykril** [comment]
@bors delegate+
r=me with the review addressed

**Pull request merged**